p2p/chain.py

```diff
@@ -695,34 +723,46 @@ def __init__(self, db: 'AsyncHeaderDB', logger: TraceLogger, token: CancelToken)
         self.logger.trace("Replying to %s with %d trie nodes", peer, len(nodes))
         peer.sub_proto.send_node_data(nodes)
 
-    async def lookup_headers(self, block_number_or_hash: Union[int, bytes], max_headers: int,
-                             skip: int, reverse: bool) -> List[BlockHeader]:
+    async def lookup_headers(self,
+                             request: HeaderRequest) -> Tuple[BlockHeader, ...]:
         """
         Lookup :max_headers: headers starting at :block_number_or_hash:, skipping :skip: items
         between each, in reverse order if :reverse: is True.
         """
-        if isinstance(block_number_or_hash, bytes):
-            try:
-                header = await self.wait(
-                    self.db.coro_get_block_header_by_hash(cast(Hash32, block_number_or_hash)))
-            except HeaderNotFound:
-                self.logger.debug(
-                    "Peer requested starting header %r that is unavailable, returning nothing",
-                    block_number_or_hash)
-                return []
-            block_number = header.block_number
-        elif isinstance(block_number_or_hash, int):
-            block_number = block_number_or_hash
+        try:
+            block_numbers = await self._get_block_numbers_for_request(request)
+        except HeaderNotFound:
+            self.logger.debug(
+                "Peer requested starting header %r that is unavailable, returning nothing",
+                request.block_number_or_hash)
+            block_numbers = tuple()  # type: ignore
+
+        headers: Tuple[BlockHeader, ...] = tuple([
+            header
+            async for header
+            in self._generate_available_headers(block_numbers)
+        ])
+        return headers
+
+    async def _get_block_numbers_for_request(self,
+                                             request: HeaderRequest) -> Tuple[BlockNumber, ...]:
+        """
+        Generate the block numbers for a given `HeaderRequest`.
+        """
+        if isinstance(request.block_number_or_hash, bytes):
+            header = await self.wait(
+                self.db.coro_get_block_header_by_hash(cast(Hash32, request.block_number_or_hash)))
+            return request.generate_block_numbers(header.block_number)
+        elif isinstance(request.block_number_or_hash, int):
+            return request.generate_block_numbers()
```

Comment:
Without having read `HeaderRequest.generate_block_numbers()`, it feels odd to see it called both with/without a block number argument. Maybe it'll make sense later on, but thought I'd mention